icu4c/source/i18n/numfmt.cpp

```diff
@@ -887,14 +887,14 @@ class NFFactory : public LocaleKeyFactory {
     virtual UObject* create(const ICUServiceKey& key, const ICUService* service, UErrorCode& status) const override
     {
         if (handlesKey(key, status)) {
-            const LocaleKey& lkey = (const LocaleKey&)key;
+            const LocaleKey& lkey = dynamic_cast<const LocaleKey&>(key);
```

Comment:
This particular line, if change to static_cast, got the following run time error during the test

```
      IntlTestNumberFormatAPI {
         NumberFormat API test {
         } OK:   NumberFormat API test 
         NumberFormatRegistration {
numfmt.cpp:890:37: runtime error: control flow integrity check for type 'icu::LocaleKey' failed during base-to-derived cast (vtable address 0x7f37c1dfa730)
0x7f37c1dfa730: note: invalid vtable
 37 7f 00 00  a0 7e cc c1 37 7f 00 00  80 7b cc c1 37 7f 00 00  a0 7b cc c1 37 7f 00 00  f0 59 cc c1
              ^ 
numfmt.cpp:890:37: note: check failed in ../../lib/libicui18n.so.73, vtable located in ../../lib/libicuuc.so.73
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior numfmt.cpp:890:37 in 

```